docs/config-file.md
# Application Configuration File

The default application configuration (when none is given) is

> **IMPORTANT:** The `nats` section is required regardless of whether `httpmq` is serving the `management` or `dataplane` API groups. When serving the `management` API group, the application will ignore settings in the `dataplane` section, and vice versa.

```yaml
---
# Core NATS configuration
nats:
  # the NATS connection URI
  server_uri: nats://127.0.0.1:4222
  # the max duration for connecting to NATS server in seconds
  connect_timeout_sec: 30
  # reconnect parameters
  reconnect:
    # the max number of reconnect attempts (-1 is unlimited)
    max_attempts: -1
    # the duration between reconnect attempts in seconds
    wait_interval_sec: 15

# Management API Group configuration
management:
  # the API endpoint config parameters for the management API server
  endpoint_config:
    # the end-point path prefix for the management APIs
    path_prefix: /
  # the HTTP API / server parameters for the management API server
  api_server:
    # HTTP server parameters
    server_config:
      # the interface the HTTP server will listen on
      listen_on: 0.0.0.0
      # the port the HTTP server will listen on
      listen_port: 3000
      # the maximum duration for reading the entire request,
      # including the body in seconds. A zero or negative
      # value means there will be no timeout.
      read_timeout_sec: 60
      # the maximum duration before timing out writes of the
      # response in seconds. A zero or negative value means
      # there will be no timeout.
      write_timeout_sec: 60
      # the maximum amount of time to wait for the next request
      # when keep-alives are enabled in seconds. If IdleTimeout
      # is zero, the value of ReadTimeout is used. If both are
      # zero, there is no timeout.
      idle_timeout_sec: 600
    # logging parameters
    logging_config:
      # the message logged marking start of a request
      start_of_request_message: "Request Starting"
      # the message logged marking end of a request
      end_of_request_message: "Request Complete"
      # the list of headers to not include in logging metadata
      do_not_log_headers:
        - "WWW-Authenticate"
        - "Authorization"
        - "Proxy-Authenticate"
        - "Proxy-Authorization"

# Dataplane API Group configuration
dataplane:
  # the API endpoint config parameters for the dataplane API server
  endpoint_config:
    # the end-point path prefix for the dataplane APIs
    path_prefix: /
  # the HTTP API / server parameters for the dataplane API server
  api_server:
    server_config:
      listen_on: 0.0.0.0
      listen_port: 3001
      read_timeout_sec: 60
      write_timeout_sec: 60
      idle_timeout_sec: 600
    logging_config:
      start_of_request_message: "Request Starting"
      end_of_request_message: "Request Complete"
      do_not_log_headers:
        - "WWW-Authenticate"
        - "Authorization"
        - "Proxy-Authenticate"
        - "Proxy-Authorization"
```

## 1. Providing Custom Configuration

The usage message for `httpmq` is

```shell
$ ./httpmq.bin -h
NAME:
   httpmq.bin - application entrypoint

USAGE:
   httpmq.bin [global options] command [command options] [arguments...]

VERSION:
   v0.2.0

DESCRIPTION:
   HTTP/2 based message broker built around NATS JetStream

COMMANDS:
   management  Run the httpmq management server
   dataplane   Run the httpmq dataplane server
   help, h     Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --json-log, -j                 Whether to log in JSON format (default: false) [$LOG_AS_JSON]
   --log-level value, -l value    Logging level: [debug info warn error] (default: warn) [$LOG_LEVEL]
   --config-file value, -c value  Application config file. Use DEFAULT if not specified. [$CONFIG_FILE]
   --help, -h                     show help (default: false)
   --version, -v                  print the version (default: false)
```

The option flag "`--config-file value`" allows the user provide a custom runtime configuration file.

## 2. Selective Configuration Override

A new custom configuration does not need to specify every single parameters seen in the default configuration. Instead, the user can provide only the fields to be replaced. As an example

```yaml
---
nats:
  server_uri: nats://dev-nats:4222

management:
  api_server:
    server_config:
      listen_on: 192.168.12.128
      listen_port: 3000

dataplane:
  api_server:
    server_config:
      listen_on: 192.168.13.128
      listen_port: 3000
```

will replace the fields specified, whereas the rest of the fields will continue to use the default values set in the application.


common/config.go
package common

import "github.com/spf13/viper"

// ===============================================================================
// NATS Related Config

// NATSReconnectConfig defines reconnect parameters
type NATSReconnectConfig struct {
	// MaxAttempts sets the max number of reconnect attempts (-1 is unlimited)
	MaxAttempts int `mapstructure:"max_attempts" json:"max_attempts" validate:"gte=-1"`
	// WaitInterval is the duration between reconnect attempts in seconds
	WaitInterval int `mapstructure:"wait_interval_sec" json:"wait_interval_sec" validate:"gte=1"`
}

// NATSConfig defines parameters for connecting to NATS server
type NATSConfig struct {
	// ServerURI is the NATS connection URI
	ServerURI string `mapstructure:"server_uri" json:"server_uri" validate:"required,uri"`
	// ConnectTimeout is the max duration for connecting to NATS server in seconds
	ConnectTimeout int `mapstructure:"connect_timeout_sec" json:"connect_timeout_sec" validate:"gte=1"`
	// Reconnect defines reconnect parameters
	Reconnect NATSReconnectConfig `mapstructure:"reconnect" json:"reconnect" validate:"required,dive"`
}

// ===============================================================================
// HTTP Related Config

// HTTPServerConfig defines the HTTP server parameters
type HTTPServerConfig struct {
	// ListenOn is the interface the HTTP server will listen on
	ListenOn string `mapstructure:"listen_on" json:"listen_on" validate:"required,ip"`
	// Port is the port the HTTP server will listen on
	Port uint16 `mapstructure:"listen_port" json:"listen_port" validate:"required,gt=0,lt=65536"`
	// ReadTimeout is the maximum duration for reading the entire
	// request, including the body in seconds. A zero or negative
	// value means there will be no timeout.
	ReadTimeout int `mapstructure:"read_timeout_sec" json:"read_timeout_sec" validate:"gte=0"`
	// WriteTimeout is the maximum duration before timing out
	// writes of the response in seconds. A zero or negative value
	// means there will be no timeout.
	WriteTimeout int `mapstructure:"write_timeout_sec" json:"write_timeout_sec" validate:"gte=0"`
	// IdleTimeout is the maximum amount of time to wait for the
	// next request when keep-alives are enabled in seconds. If
	// IdleTimeout is zero, the value of ReadTimeout is used. If
	// both are zero, there is no timeout.
	IdleTimeout int `mapstructure:"idle_timeout_sec" json:"idle_timeout_sec" validate:"gte=0"`
}

// HTTPRequestLogging defines HTTP request logging parameters
type HTTPRequestLogging struct {
	// StartOfRequestMessage is the message logged marking start of a request
	StartOfRequestMessage string `mapstructure:"start_of_request_message" json:"start_of_request_message"`
	// EndOfRequestMessage is the message logged marking end of a request
	EndOfRequestMessage string `mapstructure:"end_of_request_message" json:"end_of_request_message"`
	// DoNotLogHeaders is the list of headers to not include in logging metadata
	DoNotLogHeaders []string `mapstructure:"do_not_log_headers" json:"do_not_log_headers"`
}

// HTTPConfig defines HTTP API / server parameters
type HTTPConfig struct {
	// Server defines HTTP server parameters
	Server HTTPServerConfig `mapstructure:"server_config" json:"server_config" validate:"required,dive"`
	// Logging defines operation logging parameters
	Logging HTTPRequestLogging `mapstructure:"logging_config" json:"logging_config" validate:"required,dive"`
}

// ===============================================================================
// Management Server Related Config

// ManagementEndpointConfig defines management API endpoint config
type ManagementEndpointConfig struct {
	// PathPrefix is the end-point path prefix for the management APIs
	PathPrefix string `mapstructure:"path_prefix" json:"path_prefix" validate:"required"`
}

// ManagementServerConfig defines configuration for the management API server
type ManagementServerConfig struct {
	// HTTPSetting is the HTTP API / server parameters for the management API server
	HTTPSetting HTTPConfig `mapstructure:"api_server" json:"api_server" validate:"required,dive"`
	// Endpoints is the API endpoint config parameters for the management API server
	Endpoints ManagementEndpointConfig `mapstructure:"endpoint_config" json:"endpoint_config" validate:"required,dive"`
}

// ===============================================================================
// Dataplane Server Related Config

// DataplaneEndpointConfig defines dataplane API endpoint config
type DataplaneEndpointConfig struct {
	// PathPrefix is the end-point path prefix for the dataplane APIs
	PathPrefix string `mapstructure:"path_prefix" json:"path_prefix" validate:"required"`
}

// DataplaneServerConfig defines configuration for the dataplane API server
type DataplaneServerConfig struct {
	// HTTPSetting is the HTTP API / server parameters for the dataplane API server
	HTTPSetting HTTPConfig `mapstructure:"api_server" json:"api_server" validate:"required,dive"`
	// Endpoints is the API endpoint config parameters for the dataplane API server
	Endpoints DataplaneEndpointConfig `mapstructure:"endpoint_config" json:"endpoint_config" validate:"required,dive"`
}

// ===============================================================================
// Complete Config

// SystemConfig defines the complete system config used by either management or dataplane server
type SystemConfig struct {
	// NATS are the NATS related config parameters
	NATS NATSConfig `mapstructure:"nats" json:"nats" validate:"required,dive"`
	// Management are the management API server configs
	Management *ManagementServerConfig `mapstructure:"management,omitempty" json:"management,omitempty" validate:"omitempty,dive"`
	// Dataplane are the dataplane API server configs
	Dataplane *DataplaneServerConfig `mapstructure:"dataplane,omitempty" json:"dataplane,omitempty" validate:"omitempty,dive"`
}

// ===============================================================================

// InstallDefaultConfigValues installs default config parameters in viper
func InstallDefaultConfigValues() {
	// Default NATS settings
	viper.SetDefault("nats.server_uri", "nats://127.0.0.1:4222")
	viper.SetDefault("nats.connect_timeout_sec", 30)
	viper.SetDefault("nats.reconnect.max_attempts", -1)
	viper.SetDefault("nats.reconnect.wait_interval_sec", 15)

	// Default Management server settings
	viper.SetDefault("management.endpoint_config.path_prefix", "/")
	viper.SetDefault("management.api_server.server_config.listen_on", "0.0.0.0")
	viper.SetDefault("management.api_server.server_config.listen_port", 3000)
	viper.SetDefault("management.api_server.server_config.read_timeout_sec", 60)
	viper.SetDefault("management.api_server.server_config.write_timeout_sec", 60)
	viper.SetDefault("management.api_server.server_config.idle_timeout_sec", 600)
	viper.SetDefault(
		"management.api_server.logging_config.start_of_request_message", "Request Starting",
	)
	viper.SetDefault(
		"management.api_server.logging_config.end_of_request_message", "Request Complete",
	)
	viper.SetDefault(
		"management.api_server.logging_config.do_not_log_headers", []string{
			"WWW-Authenticate", "Authorization", "Proxy-Authenticate", "Proxy-Authorization",
		},
	)

	// Default Dataplane server settings
	viper.SetDefault("dataplane.endpoint_config.path_prefix", "/")
	viper.SetDefault("dataplane.api_server.server_config.listen_on", "0.0.0.0")
	viper.SetDefault("dataplane.api_server.server_config.listen_port", 3001)
	viper.SetDefault("dataplane.api_server.server_config.read_timeout_sec", 60)
	viper.SetDefault("dataplane.api_server.server_config.write_timeout_sec", 60)
	viper.SetDefault("dataplane.api_server.server_config.idle_timeout_sec", 600)
	viper.SetDefault(
		"dataplane.api_server.logging_config.start_of_request_message", "Request Starting",
	)
	viper.SetDefault(
		"dataplane.api_server.logging_config.end_of_request_message", "Request Complete",
	)
	viper.SetDefault(
		"dataplane.api_server.logging_config.do_not_log_headers", []string{
			"WWW-Authenticate", "Authorization", "Proxy-Authenticate", "Proxy-Authorization",
		},
	)
}
